Python/ceval.c

```diff
@@ -5221,19 +5221,28 @@ import_from(PyThreadState *tstate, PyObject *v, PyObject *name)
         pkgname_or_unknown = pkgname;
     }
 
+    _Py_IDENTIFIER(__spec__);
+    PyObject* spec = _PyObject_GetAttrId(v, &PyId___spec__);
+    Py_XINCREF(spec);
+    const char* partially_initialized =
+        _PyModuleSpec_IsInitializing(spec) ?
+        " (partially initialized module -- most likely due to a circular import)" :
+        "";
+    Py_XDECREF(spec);
+
     if (pkgpath == NULL || !PyUnicode_Check(pkgpath)) {
         _PyErr_Clear(tstate);
         errmsg = PyUnicode_FromFormat(
-            "cannot import name %R from %R (unknown location)",
-            name, pkgname_or_unknown
+            "cannot import name %R from %R%s (unknown location)",
+            name, pkgname_or_unknown, partially_initialized
         );
         /* NULL checks for errmsg and pkgname done by PyErr_SetImportError. */
         PyErr_SetImportError(errmsg, pkgname, NULL);
     }
     else {
         errmsg = PyUnicode_FromFormat(
-            "cannot import name %R from %R (%S)",
-            name, pkgname_or_unknown, pkgpath
+            "cannot import name %R from %R%s (%S)",
```

Comment:
Maybe make the error message more similar to the one of an AttributeError? `"cannot import name %R from partially initialized module %R (most likely due to a circular import)"`?

Reply:
I initially had that, but it made the code a lot harder to follow / properly refcount -- I could go back to that though

let me know if you feel strongly and I can try and make a patch for that

Reply:
No, I do not feel strongly.

But what can be hard here? You can just use different format strings:
```c
const char *msgfmt = _PyModuleSpec_IsInitializing(spec) ? "..." : "...";
...
errmsg = PyUnicode_FromFormat(msgfmt, name, pkgname_or_unknown, pkgpath);
```

Reply:
ah yeah that might work, the complication is that there's 4 different format strings flowing through this path (+/- initializing, +/- unknown path) -- I'll see what I can do 👍